Alerting: Remove vendored models in migration service

JacobsonMT opened the pull request:
Followup to #72702.

This PR replaces the vendored models in the migration with their equivalent ngalert models. It also replaces the raw SQL selects and inserts with service calls:

- Dashboard Alerts: 9f5b6b2ed2efb9f6bfeb41de7c1172938534ce1d
  - `alertRule` -> `ngmodels.AlertRule`
- Alertmanager Configuration: 9f5b6b2ed2efb9f6bfeb41de7c1172938534ce1d
  - `PostableUserConfig`, `PostableApiAlertingConfig`, `PostableApiReceiver`, `Route`, `ObjectMatchers`, `PostableGrafanaReceiver` to their `apimodels` equivalents of the same name. 
- Secure settings encryption: 149ec38d6871433c9d2d4157f05f9562d43d9cf7
  -  `util.Encrypt` ->  `secrets.Service.Encrypt`
- Legacy notification channels: 42dcb8009510eaeebb10a2a8995ae250e09551f9
  - `notificationChannel` -> `legacymodels.AlertNotification`
- Dashboards:
  - `dashboard` -> `dashboards.Dashboard` d0230235e0e2cd3ac30fe3db5acfd53669b93170
  - Raw SQL -> services a41f32dd89b939c0b4db1825c1510510a9c08f47
-  Datasources: 53b9e478cdfdef680dc8ca889bb3067263c0216d
  - `dsUIDLookup` -> `datasources.CacheService`

It also fills in some gaps in the testing suite around:

- Migration of alert rules: verifying that the actual data model (queries, conditions) are correct 9a7cfa959171e60d97061d12f7ecbb5317059485
- Secure settings migration: verifying that secure fields remain encrypted for all available notifiers and certain fields migrate from plain text to encrypted secure settings correctly e7d399368816cf06ec88712173df655340f00d3a

Replacing the checks for custom dashboard ACLs will be replaced in a separate targeted PR as it will be complex enough alone.

**NOTE**: Currently targets main as https://github.com/grafana/grafana/pull/72702 and this will be reviewed/tested together. We will still merge them separately to retain git history though.

JacobsonMT committed:
Move necessary ualert code to ngalert

JacobsonMT committed:
Refactor migration to not need migrator.Migrator

JacobsonMT committed:
Hook up new migration service to startup

Run before ngalert startup.
Use kvstore to check/save whether migration has been run.
Force run dashboard provisioning before migration.
Let dashboard provisioner provision legacy alerts even if UA is enabled.

JacobsonMT committed:
Remove condition from AddDashboardUIDPanelIDMigration, add to MigrationService

JacobsonMT committed:
Remove condition from RerunDashAlertMigration and give it a clearer name

JacobsonMT committed:
Remove condition from extractAlertmanagerConfigurationHistory

This always ran before the main migration, so it never did anything to new
legacy migrations.

JacobsonMT committed:
Remove condition from UpdateRuleGroupIndexMigration

This fixes RuleGroupIndex for existing rules to be incremental. Not necessary
to add this logic to MigrationService since each alert gets its own group.

JacobsonMT committed:
Remove CreateDefaultFoldersForAlertingMigration and other unused ones

CreateDefaultFoldersForAlertingMigration no longer has a purpose as the legacy
migration will create the general alerting folder when necessary. Also, creating
a new folder via the rule creation page is simpler than when this migration was
created.

Removes some leftover unused migration code.

JacobsonMT committed:
Fix migration tests to work with new design

JacobsonMT committed:
Remove provisioning before migration for now to keep PR targeted

JacobsonMT committed:
Move MigrationService to its own BackgroundService and revert changes to ngalert

MigrationService will no longer be guaranteed to run before ngalert starts, but
it should only cause a slight delay as the scheduler catches up after migration.

JacobsonMT committed:
Add migration to copy legacy migration status to kvstore

Also, replaces kvstore.AllOrganizations with 0 as kvstore.AllOrganizations has
special meaning when using kvstore.Get.

JacobsonMT committed:
Fix migration_test

JacobsonMT committed:
Replace vendored alertRule and PostableUserConfig with ngalert versions

JacobsonMT committed:
Add more unit tests for alert rules migration

JacobsonMT committed:
Improve testing for migration of channel secure settings

JacobsonMT committed:
Replace vendored notificationChannel with AlertNotification

JacobsonMT committed:
Encrypt migrated secureSettings with secrets.Service

JacobsonMT committed:
Replace vendored dashboard with dashboards.Dashboard

JacobsonMT committed:
Improve testing for creating new folders when migrating

JacobsonMT committed:
Replace vendored dashboard & sql with dashboards.Dashboard and services

Leaves custom ACL logic for followup as it needs to be completely rewritten for
rbac.

JacobsonMT committed:
Replace raw sql datasource lookup with datasources.CacheService

JacobsonMT committed:
Propagate session through context

JacobsonMT committed:
Add ServerLockService to support HA

JacobsonMT committed:
Fix migration_test

JacobsonMT committed:
Replace panic with error on revert without force_migration config

JacobsonMT committed:
Use logtest.Fake

JacobsonMT committed:
Skip error linting for service-level ForceMigrationError

JacobsonMT committed:
Linting

JacobsonMT committed:
Revert accidental change

JacobsonMT commented:
Changing target branch to main as https://github.com/grafana/grafana/pull/72702 and this will be reviewed/tested together. We will still merge them separately to retain git history though.

JacobsonMT committed:
Assert accesscontrol.BackgroundUser back to user.SignedInUser

JacobsonMT commented:
Updated main and rebased

JacobsonMT committed:
Fix regression: unknown datasource id errors instead of uses empty uid

yuri-tceretian commented on pkg/services/ngalert/migration/migration_test.go:
```diff
@@ -904,6 +904,23 @@ func TestDashAlertQueryMigration(t *testing.T) {
 			},
 		},
+		{
+			name: "alerts with unknown datasource id migrates with empty datasource uid",
```
Should we migrate such rules?

JacobsonMT replied:
Unclear, this functionality was part of the original migration draft from what I see. As it is now, it will migrate and show like this:

![image](https://github.com/grafana/grafana/assets/8484471/62635599-76e8-42d8-a780-189bf2a5e677)

This seems ok to me, but I don't feel strongly about it.

JacobsonMT committed:
Noop refactor - combine most external store access into migration/store

yuri-tceretian reviewed (commented):
Tremendous work!! I have tested migration from 9.5.9 to the PR with single instance and HA (3 nodes), and the migration works as expected. I could not find any problems but one:
    
  - the PR changes when the migration is run: after sqlmigration and after file provisioning. The latter is very important in this case. When I provision a file for new Grafana Alerting and turn on migration at the same time, the migration fails when my provisioned rule has title that is used by one of the migrated rules. I think this is important enough to be fixed in this PR.
  
  In my opininon, migration should clean up the Grafana alerting tables before creating new records. We already do that but when migration is rolled back. So, I think this will be the correct behavior. 
 
  Also, I found that table provenance_type is not cleaned up during rollback. This is not something missed in this particular PR, though. So, I think we can address it in a follow-up.
  
  I left some misc comments below. I think some of them could be addressed in a follow-up.
  
  Huuuggeee work. :tada:

yuri-tceretian reviewed (approved):
Most of my comments have been addressed in follow-up PRs.  LGTM

JacobsonMT committed:
Fix merge conflict with ed7d29f2b98e0e1221327a55a4096499ef78d3a2

JacobsonMT committed:
Remove vendored dashAlert

JacobsonMT committed:
Refactor migration to be per org

This is a needed for upcoming work, so we do it separately for clarity.

JacobsonMT committed:
Keep track of new folders using kvstore instead of dashboard.created_by

Setting dashboard.created_by to -8 as a way to track which folders were
created by the migration is no longer possible. Instead, we store the
newly created folders uids in the kvstore and use it on revert.

yuri-tceretian reviewed (approved):
The new changes in this PR are LGTM. The scope has changed a little bit. So, please update the description.

rwwiv reviewed (approved):
Reviewed along with @yuri-tceretian in-person, LGTM 🚀

JacobsonMT committed:
Fix linting
